src/events/message/commands/fun/rankup.ts

```diff
@@ -0,0 +1,189 @@
+  "2dan": 97225,
+  "3dan": 264305,
+  "4dan": 718477,
+  "5dan": 1953045,
+  "6dan": 5308948,
+  "7dan": 14431239,
+  "8dan": 39228196,
+};
+
+const RANKPOINTS: { [rank: string]: number } = {
+  "8kyu": 2,
+  "7kyu": 3,
+  "6kyu": 8,
+  "5kyu": 21,
+  "4kyu": 55,
+  "3kyu": 149,
+  "2kyu": 404,
+  "1kyu": 1097,
+};
+
+// Format final output string
+function formatResult(
+  user: string,
+  ranks: [number, string][],
+  target: string,
+  mode: Mode,
+  lang?: string
+) {
+  const maxLen = Math.max(...ranks.map((v) => String(v[0]).length));
+  const rankStr = ranks
+    .map(([div, rank], i) => {
+      if (!div) return "";
+      const pad = " ".repeat(maxLen - String(div).length + 1);
+      return mode === EACH
+        ? "\n" + (i ? "or " : "   ") + div + pad + rank
+        : "\n" + div + pad + rank;
+    })
+    .join("");
+  return `${user} needs to complete:
+\`\`\`
+${rankStr}
+\`\`\`
+to ${target}${lang ? " in " + lang : ""}`;
+}
+
+export type Options = {
+  mode?: string;
+  language?: string;
+  target?: string;
+  limit?: string;
+};
+
+async function getNextRank(
+  score: number,
+  user: string,
+  targ?: string,
+  lang?: string
+): Promise<[string, number] | string> {
+  // If target is a rank
+  if (targ && /[1-8](?:kyu|dan)/.test(targ)) {
+    let res: [string, number] = ["reach `" + targ + "`", RANKTHRESHOLDS[targ]];
+    if (res[1] <= score) return "Target has already been reached";
+    return res;
+  }
+
+  // If target is a user
+  if (targ) {
+    const targScore: number | string = await getUser(targ, lang).then(
+      (res) => res,
+      () => "Target user could not be found"
+    );
+    if (typeof targScore == "string") return targScore;
+    if (targScore < score) return `${user} has already reached ${targ}'s rank`;
+    return ["overtake " + targ, targScore - score + 1];
+  }
+
+  // Otherwise take next rank above user's current
+  let [nextRank, remaining] = Object.entries(RANKTHRESHOLDS)
+    .map(([rank, points]): [string, number] => [rank, points - score])
+    .filter(([, v]) => v > 0)
+    .reduce((a, c) => (a[1] < c[1] ? a : c));
+  nextRank = "reach `" + nextRank + "`";
+  return [nextRank, remaining];
+}
+
+export default async function (message: Message, args: CommandArg[], opts: Options) {
+  // Helper function
+  const send = (msg: string) => {
+    message.channel.send(msg);
+  };
+  let username: string;
+  if (args.length == 0) {
+    if (!message.member?.displayName) return;
+    username = message.member.displayName;
+  } else {
+    // Input validation
+    const result = z.tuple([word()]).safeParse(args);
+    if (!result.success) {
+      message.reply(USAGE);
+      return;
+    }
+    [username] = result.data;
+  }
+
+  // Get mode
+  const mode = isMode(opts.mode) ? opts.mode : DEFAULTMODE;
+
+  // Get user data
+  const score: number | string = await getUser(username, opts.language).then(
+    (res) => res,
+    () => `${username} is doomed to stay at :9kyu: forever. (user not found)`
```

Comment:
My logic was, the user was not found and therefore cannot progress past 8kyu. I think using callers name would be confusing, implying it checked CW for their name instead of the one provided.

I can also remove this and provide a more "normal" message, if that would be better.

Reply:
Huh, I thought the message was referring to the mistake the caller made. Like you (the caller) can't even type the username correctly.

Yeah, I prefer a normal message.